### PR TITLE
[MIST-1128] Implement automatic scale in/out policy

### LINKED ISSUE #1128: Implement automatic scale in/out policy
We need to implement dynamic scaling policy to determine whether the current cluster is overloaded or idle, and take necessary automatic scaling actions.

### PR DESCRIPTION
**Pull request by DifferentSC**
This PR closes #1128 via

* Implement the default automatic scaling in/out policy
* Add necessary interfaces & parameters
* Restructure the package of `scaling` and `allocation`

### CODE REVIEW

File: mist-core/src/main/java/edu/snu/mist/core/master/MistMaster.java [hunk 1]
```diff
@@ -118,6 +134,13 @@ private MistMaster(
       taskRequestor.recoverTaskConn();
       masterSetupFinished.setFinished();
     }
+    if (dynamicScalingEnabled) {
+      // Start dynamic scaling monitoring.
+      final ScheduledExecutorService dynamicScalingService = Executors.newSingleThreadScheduledExecutor();
+      final Injector injector = tang.newInjector();
+      final DynamicScalingPolicy policy = injector.getInstance(DefaultDynamicScalingPolicy.class);
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/master/MistMaster.java]: How about creating `DynamicScalingManager` and injecting it? And do we only consider the periodic policy? can't we implement on-demand scale in/out?

File: pom.xml [hunk 2]
```diff
@@ -149,6 +149,24 @@ limitations under the License.
                     </execution>
                 </executions>
             </plugin>
+            <plugin>
+                <groupId>org.codehaus.mojo</groupId>
```
Comment (taegeonum) [inline pom.xml]: What is this for?

File: mist-core/src/main/java/edu/snu/mist/core/parameters/DynamicScalingPeriod.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+package edu.snu.mist.core.parameters;
+
+import org.apache.reef.tang.annotations.NamedParameter;
+
+/**
+ * The monitoring period for dynamic scaling.
+ */
+@NamedParameter(doc = "The monitoring period for dynamic scaling.",
+short_name = "dynamic_scale_period", default_value = "300000")
+public class DynamicScalingPeriod {
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/parameters/DynamicScalingPeriod.java]: final

File: mist-core/src/main/java/edu/snu/mist/core/parameters/DynamicScalingEnabled.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+import org.apache.reef.tang.annotations.Name;
+import org.apache.reef.tang.annotations.NamedParameter;
+
+/**
+ * Indicates whether dynamic scale in / out is enabled or not.
+ */
+@NamedParameter(doc = "Enabling scaling in / out", default_value = "false",
+short_name = "dynamic_scale_enabled")
+public class DynamicScalingEnabled implements Name<Boolean> {
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/parameters/DynamicScalingEnabled.java]: final

File: mist-core/src/main/java/edu/snu/mist/core/master/lb/scaling/DynamicScalingRunner.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+/**
+ * The runner which runs dynamic scaling.
+ */
+public final class DynamicScalingRunner implements Runnable {
+
+  private final DynamicScalingPolicy dynamicScalingPolicy;
+
+  public DynamicScalingRunner(
+      final DynamicScalingPolicy dynamicScalingPolicy
+  ) {
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/master/lb/scaling/DynamicScalingRunner.java]: rm new line

File: mist-core/src/main/java/edu/snu/mist/core/master/lb/scaling/DefaultDynamicScalingPolicy.java [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+      final TaskStatsMap taskStatsMap,
+      @Parameter(MaxTaskNum.class) final int maxTaskNum,
+      @Parameter(MinTaskNum.class) final int minTaskNum,
+      @Parameter(IdleTaskLoadThreshold.class) final double idleTaskLoadThreshold,
+      @Parameter(OverloadedTaskLoadThreshold.class) final double overloadedTaskLoadThreshold,
+      @Parameter(ScaleInGracePeriod.class) final long scaleInGracePeriod,
+      @Parameter(ScaleOutGracePeriod.class) final long scaleOutGracePeriod,
+      @Parameter(ScaleInIdleTaskRate.class) final double scaleInIdleTaskRate,
+      @Parameter(ScaleOutOverloadedTaskRate.class) final double scaleOutOverloadedTaskRate
+  ) {
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/master/lb/scaling/DefaultDynamicScalingPolicy.java]: rm the new line

File: mist-core/src/main/java/edu/snu/mist/core/master/lb/parameters/ScaleOutOverloadedTaskRate.java [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+import org.apache.reef.tang.annotations.Name;
+import org.apache.reef.tang.annotations.NamedParameter;
+
+/**
+ * The rate of overloaded nodes.
+ */
+@NamedParameter(doc = "The overloaded cluster rate for scaling-out.", default_value = "0.8",
+short_name = "scale_out_task_rate")
+public class ScaleOutOverloadedTaskRate implements Name<Double> {
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/master/lb/parameters/ScaleOutOverloadedTaskRate.java]: final

File: mist-core/src/main/java/edu/snu/mist/core/master/lb/parameters/ScaleInIdleTaskRate.java [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+import org.apache.reef.tang.annotations.Name;
+import org.apache.reef.tang.annotations.NamedParameter;
+
+/**
+ * The rate of idle tasks for scaling-in.
+ */
+@NamedParameter(doc = "The rate of idle tasks for scaling-in", default_value = "0.5",
+short_name = "scale_in_task_rate")
+public class ScaleInIdleTaskRate implements Name<Double> {
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/master/lb/parameters/ScaleInIdleTaskRate.java]: final

File: mist-core/src/main/java/edu/snu/mist/core/master/lb/parameters/OverloadedTaskLoadThreshold.java [hunk 9]
```diff
@@ -22,5 +22,5 @@
  * The threshold for determining whether a task is overloaded or not.
  */
 @NamedParameter(doc = "The overloaded task threshold.", default_value = "0.95", short_name = "over_thd")
-public class OverloadedTaskThreshold implements Name<Double> {
+public class OverloadedTaskLoadThreshold implements Name<Double> {
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/master/lb/parameters/OverloadedTaskLoadThreshold.java]: final

File: mist-core/src/main/java/edu/snu/mist/core/master/lb/parameters/MinTaskNum.java [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+
+import org.apache.reef.tang.annotations.Name;
+import org.apache.reef.tang.annotations.NamedParameter;
+
+/**
+ * The minimum number of tasks when dynamic scaling is enabled.
+ */
+@NamedParameter(doc = "The minimum number of tasks", short_name = "min_task_num")
+public class MinTaskNum implements Name<Integer> {
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/master/lb/parameters/MinTaskNum.java]: final

File: mist-core/src/main/java/edu/snu/mist/core/master/lb/parameters/MaxTaskNum.java [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+
+import org.apache.reef.tang.annotations.Name;
+import org.apache.reef.tang.annotations.NamedParameter;
+
+/**
+ * The maximum number of tasks when dynamic scaling is enabled.
+ */
+@NamedParameter(doc = "The maxinum number of tasks when dynamic scaling is enabled",
+short_name = "max_task_num")
+public class MaxTaskNum implements Name<Integer> {
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/master/lb/parameters/MaxTaskNum.java]: final

File: pom.xml (hunk 2 above)
Comment (DifferentSC) [reply pom.xml]: This is for automatically adding avro-generated codes to IntelliJ without any configuration. I will make a separate PR for this.

File: mist-core/src/main/java/edu/snu/mist/core/master/lb/parameters/DynamicScalingPeriod.java [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+ */
+package edu.snu.mist.core.master.lb.parameters;
+
+import org.apache.reef.tang.annotations.Name;
+import org.apache.reef.tang.annotations.NamedParameter;
+
+/**
+ * The monitoring period for dynamic scaling.
+ */
+@NamedParameter(doc = "The monitoring period for dynamic scaling.",
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/master/lb/parameters/DynamicScalingPeriod.java]: (ms)?

File: mist-core/src/main/java/edu/snu/mist/core/master/lb/parameters/MinTaskNum.java [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+ * limitations under the License.
+ */
+package edu.snu.mist.core.master.lb.parameters;
+
+
+import org.apache.reef.tang.annotations.Name;
+import org.apache.reef.tang.annotations.NamedParameter;
+
+/**
+ * The minimum number of tasks when dynamic scaling is enabled.
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/master/lb/parameters/MinTaskNum.java]: is this for scaling in? if it is true, could you please modify the comment?

File: mist-core/src/main/java/edu/snu/mist/core/master/lb/parameters/MaxTaskNum.java [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+ */
+package edu.snu.mist.core.master.lb.parameters;
+
+import org.apache.reef.tang.annotations.Name;
+import org.apache.reef.tang.annotations.NamedParameter;
+
+/**
+ * The maximum number of tasks when dynamic scaling is enabled.
+ */
+@NamedParameter(doc = "The maxinum number of tasks when dynamic scaling is enabled",
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/master/lb/parameters/MaxTaskNum.java]: is this for scaling out? if it is true, could you please modify the comment?

File: mist-core/src/main/java/edu/snu/mist/core/master/lb/parameters/ScaleInGracePeriod.java [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+ */
+package edu.snu.mist.core.master.lb.parameters;
+
+import org.apache.reef.tang.annotations.Name;
+import org.apache.reef.tang.annotations.NamedParameter;
+
+/**
+ * The grace period for scaling-in.
+ */
+@NamedParameter(doc = "The grace period for scaling-in", default_value = "60000",
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/master/lb/parameters/ScaleInGracePeriod.java]: (ms)?

File: mist-core/src/main/java/edu/snu/mist/core/master/lb/parameters/ScaleOutGracePeriod.java [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+ */
+package edu.snu.mist.core.master.lb.parameters;
+
+import org.apache.reef.tang.annotations.Name;
+import org.apache.reef.tang.annotations.NamedParameter;
+
+/**
+ * The grace period for scaling out.
+ */
+@NamedParameter(doc = "The grace period for scaling out", default_value = "60000",
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/master/lb/parameters/ScaleOutGracePeriod.java]: (ms)?

File: mist-core/src/main/java/edu/snu/mist/core/master/lb/scaling/PeriodicDynamicScalingManager.java [hunk 17]
```diff
@@ -0,0 +1,215 @@
...
+      @Parameter(DynamicScalingPeriod.class) final long dynamicScalingPeriod,
+      @Parameter(MaxTaskNum.class) final int maxTaskNum,
+      @Parameter(MinTaskNum.class) final int minTaskNum,
+      @Parameter(IdleTaskLoadThreshold.class) final double idleTaskLoadThreshold,
+      @Parameter(OverloadedTaskLoadThreshold.class) final double overloadedTaskLoadThreshold,
+      @Parameter(ScaleInGracePeriod.class) final long scaleInGracePeriod,
+      @Parameter(ScaleOutGracePeriod.class) final long scaleOutGracePeriod,
+      @Parameter(ScaleInIdleTaskRate.class) final double scaleInIdleTaskRate,
+      @Parameter(ScaleOutOverloadedTaskRate.class) final double scaleOutOverloadedTaskRate
+  ) {
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/master/lb/scaling/PeriodicDynamicScalingManager.java]: rm the new line

File: mist-core/src/main/java/edu/snu/mist/core/master/lb/scaling/PeriodicDynamicScalingManager.java [hunk 18]
```diff
@@ -0,0 +1,215 @@
...
+  }
+
+  @Override
+  public void startAutoScaling() {
+    scheduledExecutorService.schedule(new AutoScaleRunner(), dynamicScalingPeriod, TimeUnit.MILLISECONDS);
+  }
+
+  @Override
+  public void close() throws Exception {
+    scheduledExecutorService.shutdown();
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/master/lb/scaling/PeriodicDynamicScalingManager.java]: scheduledExecutorService.shutdown();
scheduledExecutorService.awaitTermination(..);

File: mist-core/src/main/java/edu/snu/mist/core/master/lb/scaling/PeriodicDynamicScalingManager.java [hunk 19]
```diff
@@ -0,0 +1,215 @@
...
+    this.minTaskNum = minTaskNum;
+    this.idleTaskLoadThreshold = idleTaskLoadThreshold;
+    this.overloadedTaskLoadThreshold = overloadedTaskLoadThreshold;
+    this.scaleInGracePeriod = scaleInGracePeriod;
+    this.scaleOutGracePeriod = scaleOutGracePeriod;
+    this.scaleInIdleTaskRate = scaleInIdleTaskRate;
+    this.scaleOutOverloadedTaskRate = scaleOutOverloadedTaskRate;
+    this.idleTimeElapsed = 0L;
+    this.overloadedTimeElapsed = 0L;
+    this.lastMeasuredTimestamp = 0L;
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/master/lb/scaling/PeriodicDynamicScalingManager.java]: = System.currentTimeMillies();

File: mist-core/src/main/java/edu/snu/mist/core/master/lb/scaling/PeriodicDynamicScalingManager.java [hunk 20]
```diff
@@ -0,0 +1,215 @@
...
+  private final class AutoScaleRunner implements Runnable {
+
+    private AutoScaleRunner() {
+      // Do nothing.
+    }
+
+    @Override
+    public void run() {
+      final long oldTimeStamp;
+      if (lastMeasuredTimestamp == 0L) {
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/master/lb/scaling/PeriodicDynamicScalingManager.java]: We don't need this `if else` if we set the lastMeasuredTimestamp in the constructor

File: mist-core/src/main/java/edu/snu/mist/core/master/lb/scaling/PeriodicDynamicScalingManager.java [hunk 21]
```diff
@@ -0,0 +1,215 @@
...
+        overloadedTimeElapsed = 0;
+      }
+
+      if (clusterIdle) {
+        idleTimeElapsed += lastMeasuredTimestamp - oldTimeStamp;
+      } else {
+        idleTimeElapsed = 0;
+      }
+
+      if (overloadedTimeElapsed > scaleOutGracePeriod && taskStatsMap.getTaskList().size() < maxTaskNum) {
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/master/lb/scaling/PeriodicDynamicScalingManager.java]: please add this `if` within line 185

File: mist-core/src/main/java/edu/snu/mist/core/master/lb/scaling/PeriodicDynamicScalingManager.java [hunk 22]
```diff
@@ -0,0 +1,215 @@
...
+        // TODO: [MIST-1130] Perform automatic scale-out.
+        overloadedTimeElapsed = 0;
+      } else if (idleTimeElapsed > scaleInGracePeriod && taskStatsMap.getTaskList().size() > minTaskNum) {
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/master/lb/scaling/PeriodicDynamicScalingManager.java]: please add this `else if` within line 191

File: mist-core/src/main/java/edu/snu/mist/core/master/lb/scaling/PeriodicDynamicScalingManager.java [hunk 23]
```diff
@@ -0,0 +1,215 @@
...
+  }
+
+  private boolean isClusterOverloaded() {
+    int overloadedTaskNum = 0;
+    for (final Map.Entry<String, TaskStats> entry : taskStatsMap.entrySet()) {
+      if (entry.getValue().getTaskLoad() > overloadedTaskLoadThreshold) {
+        overloadedTaskNum += 1;
+      }
+    }
+    final double overloadedTaskRate = overloadedTaskNum / taskStatsMap.getTaskList().size();
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/master/lb/scaling/PeriodicDynamicScalingManager.java]: rate -> ratio?

File: mist-core/src/main/java/edu/snu/mist/core/master/lb/parameters/ScaleOutOverloadedTaskRate.java [hunk 24]
```diff
@@ -0,0 +1,27 @@
...
+
+import org.apache.reef.tang.annotations.Name;
+import org.apache.reef.tang.annotations.NamedParameter;
+
+/**
+ * The rate of overloaded nodes.
+ */
+@NamedParameter(doc = "The overloaded cluster rate for scaling-out.", default_value = "0.8",
+short_name = "scale_out_task_rate")
+public final class ScaleOutOverloadedTaskRate implements Name<Double> {
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/master/lb/parameters/ScaleOutOverloadedTaskRate.java]: rate -> ratio?

File: mist-core/src/main/java/edu/snu/mist/core/master/lb/parameters/ScaleInIdleTaskRate.java [hunk 25]
```diff
@@ -0,0 +1,27 @@
...
+
+import org.apache.reef.tang.annotations.Name;
+import org.apache.reef.tang.annotations.NamedParameter;
+
+/**
+ * The rate of idle tasks for scaling-in.
+ */
+@NamedParameter(doc = "The rate of idle tasks for scaling-in", default_value = "0.5",
+short_name = "scale_in_task_rate")
+public final class ScaleInIdleTaskRate implements Name<Double> {
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/master/lb/parameters/ScaleInIdleTaskRate.java]: rate -> ratio?

File: mist-core/src/main/java/edu/snu/mist/core/master/lb/scaling/PeriodicDynamicScalingManager.java (hunk 18 above)
Comment (DifferentSC) [reply mist-core/src/main/java/edu/snu/mist/core/master/lb/scaling/PeriodicDynamicScalingManager.java]: Addressed.

**Comment by taegeonum** [comment]
LGTM. I'm merging it

**Pull request merged**